wrapper/src/main/java/software/amazon/jdbc/plugin/failover2/FailoverConnectionPlugin.java

```diff
@@ -0,0 +1,619 @@
+import software.amazon.jdbc.AwsWrapperProperty;
+import software.amazon.jdbc.HostListProviderService;
+import software.amazon.jdbc.HostRole;
+import software.amazon.jdbc.HostSpec;
+import software.amazon.jdbc.JdbcCallable;
+import software.amazon.jdbc.PluginManagerService;
+import software.amazon.jdbc.PluginService;
+import software.amazon.jdbc.PropertyDefinition;
+import software.amazon.jdbc.hostavailability.HostAvailability;
+import software.amazon.jdbc.plugin.AbstractConnectionPlugin;
+import software.amazon.jdbc.plugin.failover.FailoverFailedSQLException;
+import software.amazon.jdbc.plugin.failover.FailoverMode;
+import software.amazon.jdbc.plugin.failover.FailoverSuccessSQLException;
+import software.amazon.jdbc.plugin.failover.TransactionStateUnknownSQLException;
+import software.amazon.jdbc.plugin.staledns.AuroraStaleDnsHelper;
+import software.amazon.jdbc.targetdriverdialect.TargetDriverDialect;
+import software.amazon.jdbc.util.Messages;
+import software.amazon.jdbc.util.RdsUrlType;
+import software.amazon.jdbc.util.RdsUtils;
+import software.amazon.jdbc.util.SqlState;
+import software.amazon.jdbc.util.SubscribedMethodHelper;
+import software.amazon.jdbc.util.Utils;
+import software.amazon.jdbc.util.WrapperUtils;
+import software.amazon.jdbc.util.telemetry.TelemetryContext;
+import software.amazon.jdbc.util.telemetry.TelemetryCounter;
+import software.amazon.jdbc.util.telemetry.TelemetryFactory;
+import software.amazon.jdbc.util.telemetry.TelemetryTraceLevel;
+
+/**
+ * Failover Plugin v.2
+ * This plugin provides cluster-aware failover features. The plugin switches connections upon
+ * detecting communication related exceptions and/or cluster topology changes.
+ */
+public class FailoverConnectionPlugin extends AbstractConnectionPlugin {
+
+  private static final Logger LOGGER = Logger.getLogger(FailoverConnectionPlugin.class.getName());
+  private static final String TELEMETRY_WRITER_FAILOVER = "failover to writer node";
+  private static final String TELEMETRY_READER_FAILOVER = "failover to replica";
+
+  public static final AwsWrapperProperty FAILOVER_TIMEOUT_MS =
+      new AwsWrapperProperty(
+          "failoverTimeoutMs",
+          "300000",
+          "Maximum allowed time for the failover process.");
+
+  public static final AwsWrapperProperty FAILOVER_MODE =
+      new AwsWrapperProperty(
+          "failoverMode", null,
+          "Set node role to follow during failover.");
+
+  public static final AwsWrapperProperty TELEMETRY_FAILOVER_ADDITIONAL_TOP_TRACE =
+      new AwsWrapperProperty(
+          "telemetryFailoverAdditionalTopTrace", "false",
+          "Post an additional top-level trace for failover process.");
+
+  public static final AwsWrapperProperty FAILOVER_READER_HOST_SELECTOR_STRATEGY =
+      new AwsWrapperProperty(
+          "failoverReaderHostSelectorStrategy",
+          "random",
+          "The strategy that should be used to select a new reader host while opening a new connection.");
+
```

Comment:
We might need to add a couple missing properties:
- enableClusterAwareFailover, which is used in failover 1. Unless we are deprecating that property and assuming that if the user included the plugin they want failover enabled
- enableConnectFailover, which was added recently to address a customer issue. We can add this in a separate PR though

Reply:
I feel like `enableClusterAwareFailover` not utilized at all so I think we can not implement it in `failover2`. If it's important for users, we can add it easily.

As for `enableConnectFailover`, I agree. I'd like to add it as a separate PR.